File: CSharp/src/Runtime/Scripts/XFile/manifest.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Generic;

namespace FrameworkEase.Universal.X.Utility
{
    public static partial class XFile
    {
        public class FileInfo
        {
            public string Name;

            public string Hash;

            public long Size;
        }

        public class DiffInfo
        {
            public readonly List<FileInfo> Added = new();

            public readonly List<FileInfo> Modified = new();

            public readonly List<FileInfo> Deleted = new();
        }

        public class Manifest
        {
            public readonly List<FileInfo> Files = new();

            public virtual DiffInfo Compare(Manifest other)
            {
                var diffInfo = new DiffInfo();
                var selfFiles = Files;
                var otherFiles = other.Files;
                var visited = new List<FileInfo>();
                for (var i = 0; i < selfFiles.Count; i++)
                {
                    var sf = selfFiles[i];
                    var sig = false;
                    for (var j = 0; j < otherFiles.Count; j++)
                    {
                        var of = otherFiles[j];
                        if (of.Name == sf.Name)
                        {
                            if (sf.Hash != of.Hash)
                            {
                                diffInfo.Modified.Add(of);
                            }
                            sig = true;
                            visited.Add(of);
                            break;
                        }
                    }
                    if (!sig) diffInfo.Deleted.Add(sf);
                }
                for (var i = 0; i < otherFiles.Count; i++)
                {
                    var fi = otherFiles[i];
                    if (!visited.Contains(fi)) diffInfo.Added.Add(fi);
                }
                return diffInfo;
            }

            public string Stringify(bool pretty = false) { return JSON.Stringify(Files, pretty); }

            public virtual string Parse(string content)
            {
                if (string.IsNullOrEmpty(content)) return "null content for parsing mainfest";
                else
                {
                    try
                    {
                        var files = JSON.Parse<List<FileInfo>>(content);
                        if (files == null) return "invalid JSON format for parsing manifest";
                        Files.Clear();
                        Files.AddRange(files);
                    }
                    catch (Exception e)
                    {
                        XLog.Error(e);
                        return e.Message;
                    }
                }
                return string.Empty;
            }
        }
    }
}

File: Go/XFile/manifest.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XFile

import (
	"encoding/json"
	"fmt"
)

type FileInfo struct {
	Name string
	Hash string
	Size int64
}

type DiffInfo struct {
	Added    []*FileInfo
	Modified []*FileInfo
	Deleted  []*FileInfo
}

type Manifest struct {
	Files []*FileInfo
}

func (m *Manifest) Compare(other *Manifest) *DiffInfo {
	diffInfo := &DiffInfo{
		Added:    make([]*FileInfo, 0),
		Modified: make([]*FileInfo, 0),
		Deleted:  make([]*FileInfo, 0),
	}

	selfFiles := m.Files
	otherFiles := other.Files
	visited := make(map[int]bool)

	// 检查当前清单中的文件
	for i := range selfFiles {
		sf := selfFiles[i]
		found := false
		for j := range otherFiles {
			of := otherFiles[j]
			if of.Name == sf.Name {
				if sf.Hash != of.Hash {
					diffInfo.Modified = append(diffInfo.Modified, of)
				}
				found = true
				visited[j] = true
				break
			}
		}
		if !found {
			diffInfo.Deleted = append(diffInfo.Deleted, sf)
		}
	}

	// 检查目标清单中的新文件
	for i := range otherFiles {
		if !visited[i] {
			diffInfo.Added = append(diffInfo.Added, otherFiles[i])
		}
	}

	return diffInfo
}

func (m *Manifest) Stringify(pretty ...bool) string {
	if len(pretty) > 0 && pretty[0] {
		json, err := json.MarshalIndent(m.Files, "", "  ")
		if err != nil {
			fmt.Printf("XFile.Manifest.Stringify: marshal error: %v\n", err)
		}
		return string(json)
	} else {
		json, err := json.Marshal(m.Files)
		if err != nil {
			fmt.Printf("XFile.Manifest.Stringify: marshal error: %v\n", err)
		}
		return string(json)
	}
}

func (m *Manifest) Parse(content []byte) error {
	if len(content) == 0 {
		return fmt.Errorf("null content for parsing manifest")
	}
	var files []*FileInfo
	if err := json.Unmarshal(content, &files); err != nil {
		return fmt.Errorf("invalid JSON format for parsing manifest: %w", err)
	}
	m.Files = files
	return nil
}
